docs/public-api.md

```diff
@@ -0,0 +1,65 @@
+
+  Examples:
+    - ✅ `extension MyHandler: ChannelHandler { ... }`
+    - ❌ `extension Array: EventLoopGroup where Element: EventLoop { ... }`, as `Array` lives in the standard library
+
+- it is acceptable to add extensions to NIO types that either use one of your types as a non-default argument or are prefixed in a way that prevents clashes
+
+  What's the point of these restrictions? NIO might later add a member function which might then clash with yours and might stop your code from compiling and/or working.
+
+  Examples:
+    - ✅ `extension ByteBuffer { mutating func readMyType(at: Int) -> MyType }`, acceptable because it returns a `MyType`
+    - ✅ `extension ByteBuffer { mutating func myProjectReadInteger(at: Int) -> Int`, acceptable because prefixed with `myProject`
+    - ❌ `extension ByteBuffer { mutating func readBool(at: Int) -> Bool }`, because `Bool` is a standard library type not in your control
+
+### Promises the NIO Team Make
+
+Before releasing a new major version, ie. SwiftNIO 3.0.0 we promise the following:
+
+- no global types/modules/functions will be added that do not have `NIO*`/`nio*`/`CNIO*`/`cnio*`/`c_nio*` prefixes, this includes C functions that are not in the public API.
+  Examples:
+    - we might add a new global type called `NIOJetPack`
+    - we will not add a global new type called `Tomato`
+    - we might add a new module called `NIOHelpers`
+    - we will not add a new module called `Helpers`
+    - we might add a new global functions called `nioRun` (very unlikely) or `c_nio_SSL_write` (more likely)
+    - we will not add a new global function called `SSL_write`
+
+### What's the Point of Specifying All This?
+
+We believe that it's important for the ecosystem that all parties can adopt each others new versions quickly and easily. Therefore we always strongly recommend to depend on all the projects in the NIO family which have a `major version >= 1` up to the next _major_ version. By following the rules set out above we should all not run into any trouble even when run with newer versions.
+
+Example:
+
+    .package(url: "https://github.com/apple/swift-nio.git", from: "2.0.0"),
+    
+Needless to say if you require at least version `2.3.4` you would specify `from: "2.3.4"`. The important part is that you use `from: "2.3.4"` or `.upToNextMajor(from: "2.3.4")` and not `.exact` or `.upToNextMinor`.
+
+### What Happens if You Ignore this Recommendation?
+
+Hopefully, nothing at all ☺️. We have a source compatibility suite to which you can [ask to be added](https://forums.swift.org/t/register-as-swiftnio-user-to-get-ahead-of-time-security-notifications-be-added-to-the-source-compatibility-suite/17792) and we try not to break you (_within reason_). But it's impossible for us to test all of our users projects' and we don't want to lose the ability to move fast without breaking things. Certain failures like clashing protocol conformances can have delicate failure modes.
```

Comment:
While that's true and cool, the first sentence sounds a bit too soft to me... Perhaps start re stating that ignoring those rules may cause you to have to rewrite things etc, and then mention the happy news that there is a compat suite so you'd get heads up (but still MAY need to change things if team decides to have to change such internal API etc)